qiskit_experiments/data_processing/nodes.py

```diff
@@ -0,0 +1,181 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Different data analysis steps."""
+
+from abc import abstractmethod
+from typing import Any, Dict, Optional, Tuple
+import numpy as np
+
+from qiskit_experiments.data_processing.data_action import DataAction
+from qiskit_experiments.data_processing.exceptions import DataProcessorError
+
+
+class IQPart(DataAction):
+    """Abstract class for IQ data post-processing."""
+
+    def __init__(self, scale: Optional[float] = None):
+        """
+        Args:
+            scale: float with which to multiply the IQ data.
+        """
+        self.scale = scale
+        super().__init__()
+
+    @abstractmethod
+    def _process_iq(self, datum: np.array) -> np.array:
+        """Defines how the IQ point will be processed.
+
+        Args:
+            datum: a 3D array of shots, qubits, and a complex IQ point as [real, imaginary].
+
+        Returns:
+            Processed IQ point.
+        """
+
+    def _check_data_format(self, datum: Any) -> Any:
+        """Check that the IQ data has the correct format.
+
+        Args:
+            datum: A single item of data which corresponds to single-shot IQ data. It should
+                have dimension three: shots, qubits, iq-point as [real, imaginary].
+
+        Returns:
+            datum: as a numpy array.
+
+        Raises:
+            DataProcessorError: if the datum does not have the correct format.
+        """
+        if not isinstance(datum, (list, np.ndarray)):
```

Comment:
You could just do
```python
datum = np.asarray(datum, dtype=complex)  # replace dtype with what dtype its supposed to be
```
right off the bat rather than check its a list or array first.

Reply:
Done here: 597d60c